### ISSUE TITLE
Add support for slicing an edge

### ISSUE THREAD

**Issue by Thenerdstation** [google/TensorNetwork#260]
Slicing an edge with something like

```python
net.slice_edge(edge, start_index, length)
```
Would be very useful for doing parallel contractions using JAX's `pmap`. Each slice of the edge would run on a different device and the final result would just be the sum of all of the slices, which is very easy to do in jax with `psum`.

**Comment by amilsted** [comment]
Sounds good! I assume this is intended to be used on internal edges? Could you explain in a little more detail what the arguments mean?

**Comment by anak4569** [comment]
@Thenerdstation I'm interested in working on this issue; can you provide some more context on the arguments/intended behavior?

**Comment by Thenerdstation** [comment]
Certainly!

So sometimes you may want to contract your tensornetwork in parts. (For example, the "tensor cuts" they did in the [quantum supremacy benchmark ](https://arxiv.org/abs/1905.00444)), so to do that you'll want to take a "slice" of an edge.

For a basic example, let's look at a basic matmul case.

```python
a = tn.Node(np.ones((1000, 1000)))
b = tn.Node(np.ones((1000, 1000)))
edge = a[1] ^ b[0]
```

So in this case, `edge.dimension == 1000`. However, let's say we want to contract this on multiple machines in parallel (say, using `jax.pmap`). What we could do is then "slice" this edge into pieces and contract each piece separately on each device. What that would look like is then.


```
new_edge = tn.slice_edge(edge, start, 20)
```

So now that edge would have dimension 20 instead of 1000, and you could change the start position for each device you ran on (so `start` could equal 0 on one device, 20 on the next, 40 on the one afterwards, etc). 

After running `slice_edge`. the shape of `a` would be `(1000, 20)` instead of `(1000, 1000)`, and similarly `b.shape == (20, 1000)` 

Underneath, you would need to slice the underlying tensors. which would look like

```
new_a_tensor = a_tensor[:,start:length]
```
(Be careful about which axis you slice, this is just the case for this specific `a`). You likely would need to add a new `slice` method to the `BaseBackend`.

Hope that helps! Feel free to ask more questions on here if you get stuck.

**Comment by anak4569** [comment]
Question about the motivation for slice_edge: how would you coordinate a contraction using this method on multiple machines? Specifically, do we want the original `a` to change shape (which means you could lose access to the original tensor and make coordinating the contraction more difficult without (potentially?) expensive copying) or do we just want the edge's internal reference to be changed?
